### PR TITLE
fix: [filter_box] fix 2 issues in single value filter_box

### PR DESCRIPTION
**Pull request by graceguo-supercat**
### SUMMARY
This PR is to fix 2 issues in **single value** filter_box.

### BEFORE

1. Issue 1: Clear filter_box won't clear filter's default value
gGtwLBI2qw

1. Issue 2: Set default value for a single value filter_box, the column's data type is FLOAT:
Screen Shot 2020-05-17 at 6 48 57 PM
In dashboard, this string value doesn't match with filter's data type, so that it looks the same option show twice:
Screen Shot 2020-05-17 at 6 50 46 PM

Proposed solution:

1. Issue 1: When user clear filter_box, the option value should be null. While when filter_box value is undefined, should use defaultValue.
2. Issue 2: When user update defaultValue for filter_box, add type cast for numeric and boolean type column.
 

### TEST PLAN
CI and manual test.

@mistercrunch @ktmud @etr2460

### CODE REVIEW

File: superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx [hunk 1]
```diff
@@ -60,7 +63,35 @@ export default class FilterBoxItemControl extends React.Component {
...
+    const { column: selectedColumnName, multiple } = this.state;
+    if (value && !multiple && attr === 'defaultValue') {
+      // if single value filter_box,
+      // convert input value string to the column's data type
+      const { datasource } = this.props;
+      const selectedColumn = (
+        datasource.columns.filter(
+          col => col.column_name === selectedColumnName,
+        ) || []
+      ).pop();
```
Comment (ktmud) [inline superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx:76]:
```js
selectedColumn = datasource.columns.find(x => x.column_name === selectedColumnName);
```

File: superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx [hunk 2]
```diff
@@ -27,6 +27,9 @@ import SelectControl from './SelectControl';
 import CheckboxControl from './CheckboxControl';
 import TextControl from './TextControl';
 
+const INTEGRAL_TYPES = ['TINYINT', 'SMALLINT', 'INT', 'INTEGER', 'BIGINT'];
+const DECIMAL_TYPES = ['FLOAT', 'DOUBLE'];
```
Comment (ktmud) [inline superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx:31]: Can we use `Set`? It's faster than arrays.

```
const DECIMAL_TYPES = new Set(['FLOAT', 'DOUBLE'])

DECIMAL_TYPES.has(columnType)
```

Not sure what types are returned to the frontend, but are we sure these types are exhaustive: https://github.com/apache/incubator-superset/blob/a6cedaaa879348aca49a520793bb20e63d152a1f/superset/connectors/base/models.py#L480-L493

Might want to add `LONG` to `INTEGRAL_TYPES`; and `NUMERIC`, `NUMBER`, `REAL` and `DECIMAL` to `DECIMAL_TYPES`, too?

**Review by ktmud — APPROVED** [review]

**Comment by ktmud** [comment]
LGTM, just a couple of style nits.

File: superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx [hunk 3]
```diff
@@ -60,7 +63,35 @@ export default class FilterBoxItemControl extends React.Component {
     this.props.onChange(this.state);
   }
   onControlChange(attr, value) {
-    this.setState({ [attr]: value }, this.onChange);
+    let typedValue = value;
+    const { column: selectedColumnName, multiple } = this.state;
+    if (value && !multiple && attr === 'defaultValue') {
```
Comment (etr2460) [inline superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx:68]: should `defaultValue` be a string constant somewhere that's used here and in the filter config?

File: superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx [hunk 4]
```diff
@@ -60,7 +63,35 @@ export default class FilterBoxItemControl extends React.Component {
...
+        const type = selectedColumn.type.toUpperCase();
+        try {
+          if (type === 'BOOLEAN') {
+            typedValue = value === 'true';
+          } else if (INTEGRAL_TYPES.includes(type)) {
+            typedValue = parseInt(value, 10);
+          } else if (DECIMAL_TYPES.includes(type)) {
+            typedValue = parseFloat(value);
+          }
+        } catch (ex) {
```
Comment (etr2460) [inline superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx:88]: when does this fail? I don't think `parseInt` or `parseFloat` throw errors with invalid input. instead they return `NaN`. I think we can remove the try/catch block

**Review by etr2460 — APPROVED** [review]
lgtm as well after addressing Jesse's comment and mine about the try/catch

File: superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx (hunk 3 above)
Comment (graceguo-supercat) [reply superset-frontend/src/explore/components/controls/FilterBoxItemControl.jsx]: added 2 constants, `defaultValue` and `multi` (used by this feature). but there are quite a few other attributes used in many places...

**Pull request merged**